File: 32.js
function minWindow(s, t) {
    const need = new Map();
    for (let ch of t) need.set(ch, (need.get(ch) || 0) + 1);

    let missing = t.length;
    let left = 0, start = 0, minLen = Infinity;

    for (let right = 0; right < s.length; right++) {
        const ch = s[right];
        if (need.get(ch) > 0) missing--;
        need.set(ch, (need.get(ch) || 0) - 1);

        while (missing === 0) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                start = left;
            }

            const leftChar = s[left];
            need.set(leftChar, (need.get(leftChar) || 0) + 1);
            if (need.get(leftChar) > 0) missing++;
            left++;
        }
    }
    return minLen === Infinity ? "" : s.slice(start, start + minLen);
}

File: 32.py
from collections import Counter

def minWindow(s: str, t: str) -> str:
    need = Counter(t)
    missing = len(t)
    left = 0
    start = 0
    min_len = float("inf")

    for right, ch in enumerate(s):
        if need[ch] > 0:
            missing -= 1
        need[ch] -= 1

        while missing == 0:
            if right - left + 1 < min_len:
                min_len = right - left + 1
                start = left

            need[s[left]] += 1
            if need[s[left]] > 0:
                missing += 1
            left += 1

    return "" if min_len == float("inf") else s[start:start+min_len]
